Set default clipBehavior to Clip.none and use InkWell's own clipping

liyuqian opened the pull request:
This reverts https://github.com/flutter/flutter/pull/20578 and fixes https://github.com/flutter/flutter/issues/20483 by letting InkWell do its own clipping.

PathOp.intersect is not used because we have too many unit tests that rely on clipping (e.g., `paints..clipXXX()..drawCircle()`)

A golden test is also added for https://github.com/flutter/flutter/issues/20483

liyuqian committed:
Use PathOp.intersect for InkWell clipping

liyuqian committed:
Revert "Revert to defaultClipBehavior (#20578)"

This reverts commit 8de0e15e4fdd3a1be8878394a7f8c8bda948e59b.

Hixie commented on packages/flutter/lib/src/material/ink_highlight.dart:
```diff
@@ -40,13 +40,13 @@ class InkHighlight extends InteractiveInkFeature {
     @required Color color,
     BoxShape shape = BoxShape.rectangle,
-    BorderRadius borderRadius,
+    ShapeBorder border,
```
we can't remove borderRadius, that's a breaking change. We should support it by implying a rounded rectangle shape of the given border radius.

Hixie commented on packages/flutter/lib/src/material/ink_ripple.dart:
```diff
@@ -47,7 +47,7 @@ class _InkRippleFactory extends InteractiveInkFeatureFactory {
     bool containedInkWell = false,
     RectCallback rectCallback,
-    BorderRadius borderRadius,
+    ShapeBorder border,
```
ditto

Hixie commented on packages/flutter/lib/src/material/ink_well.dart:
```diff
@@ -94,7 +94,7 @@ abstract class InteractiveInkFeatureFactory {
     bool containedInkWell = false,
     RectCallback rectCallback,
-    BorderRadius borderRadius,
+    ShapeBorder border,
```
ditto

Hixie commented on packages/flutter/lib/src/material/ink_well.dart:
```diff
@@ -625,7 +623,7 @@ class InkWell extends InkResponse {
     InteractiveInkFeatureFactory splashFactory,
     double radius,
-    BorderRadius borderRadius,
+    ShapeBorder border,
```
ditto

Hixie commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -462,7 +462,7 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
       containedInkWell: true,
       // TODO(hansmuller): splash clip borderRadius should match the input decorator's border.
-      borderRadius: BorderRadius.zero,
+      border: null, // a null border generates a rectangle border clip
```
as far as i can tell it generates no clip.

Hixie replied:
oh i see, that was for the highlight. never mind.

Hixie commented:
we need some tests for the splash behaviour, so that we don't regress that again. :-)
(probably golden tests)

liyuqian committed:
Add borderRadius back to InkWell

liyuqian committed:
Use clip intead of PathOperation.intersect and update unit tests

PathOp.intersect is not used because we have too many unit tests that
rely on clipping (e.g., `paints..clipXXX()..drawCircle()`)

liyuqian committed:
Add golden test for clipping

liyuqian committed:
Rename border to customBorder

So it's clearer that it has nothing to do with borderRadius

HansMuller commented on packages/flutter/test/material/ink_paint_test.dart:
```diff
@@ -86,93 +90,49 @@ void main() {
+          if (offsetsAreClose(center, expectedCenter) && radiiAreClose(radius, expectedRadius) && paint.color.alpha == expectedAlpha)
+            return true;
+          throw '''
+          Expected: center == $expectedCenter, radius == $expectedRadius, alpha == $expectedAlpha
```
indent the text

liyuqian committed:
Update goldens version

HansMuller commented on packages/flutter/test/material/ink_paint_test.dart:
```diff
@@ -86,93 +90,49 @@ void main() {
-        Expected: center == $tapDownOffset, radius == 30.0, alpha == 0
-        Found: center == $center radius == $radius alpha == ${paint.color.alpha}''';
-    }));
+    expect(box, ripplePattern(Offset.zero, 30.0, 0));
```
tapDownOffset instead of Offset.zero (per the comment on line 112)

HansMuller commented on packages/flutter/test/material/ink_paint_test.dart:
```diff
@@ -86,93 +90,49 @@ void main() {
     bool offsetsAreClose(Offset a, Offset b) => (a - b).distance < 1.0;
     bool radiiAreClose(double a, double b) => (a - b).abs() < 1.0;
 
+    PaintPattern ripplePattern(Offset expectedCenter, double expectedRadius, int expectedAlpha) {
```
NICE

liyuqian commented:
@Hixie golden tests added. Although `PathOperation.intersect` is faster than clipping in theory, we currently have too many unit tests that rely on clipping. Maybe the small performance gain isn't worth for a big unit tests refactor for now. We can probably reconsider that in November.

liyuqian committed:
Nits

jonahwilliams commented:
I would strongly consider that any changes to clipping be followed up with a full gallery run through

liyuqian commented:
As this PR grows to be much larger than I originally expected, I splitted it into
- https://github.com/flutter/flutter/pull/20750
- https://github.com/flutter/flutter/pull/20751
- https://github.com/flutter/flutter/pull/20752

Hopefully that makes it a little easier to review, and maybe revert if something goes wrong.

liyuqian commented:
@Hixie @jonahwilliams @HansMuller Would you like to review and land this big PR or review and land 3 smaller PRs #20750, #20751, #20752 ?

jonahwilliams commented:
@liyuqian I would rather fight 100 duck sized horses

liyuqian commented:
The PR has been split into smaller PRs.
